processing/src/test/java/org/apache/druid/segment/join/PostJoinCursorTest.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    private static class CursorNoAdvance implements Cursor
+    {
+      Cursor cursor;
+
+      public CursorNoAdvance(Cursor cursor)
+      {
+        this.cursor = cursor;
+      }
+
+      @Override
+      public ColumnSelectorFactory getColumnSelectorFactory()
+      {
+        return cursor.getColumnSelectorFactory();
+      }
+
+      @Override
+      public DateTime getTime()
+      {
+        return cursor.getTime();
+      }
+
+      @Override
+      public void advance()
+      {
+        // Do nothing to simulate infinite rows
+      }
+
+      @Override
+      public void advanceUninterruptibly()
+      {
+        // Do nothing to simulate infinite rows
+      }
+
+      @Override
+      public boolean isDone()
+      {
+        return cursor.isDone();
+      }
+
+      @Override
+      public boolean isDoneOrInterrupted()
+      {
+        return cursor.isDoneOrInterrupted();
+      }
+
+      @Override
+      public void reset()
+      {
+
+      }
+    }
+  }
+
+  private static class ExceptionHandler implements Thread.UncaughtExceptionHandler
+  {
+
+    Throwable exception;
+
+    @Override
+    public void uncaughtException(Thread t, Throwable e)
+    {
+      exception = e;
+    }
+
+    public Throwable getException()
+    {
+      return exception;
+    }
+  }
+
+  @Test
+  public void testAdvanceWithInterruption()
+  {
+    Thread joinCursorThread = new Thread(() -> makeCursorAndAdvance());
+    ExceptionHandler exceptionHandler = new ExceptionHandler();
+    joinCursorThread.setUncaughtExceptionHandler(exceptionHandler);
+    joinCursorThread.start();
+    try {
+      sleep(5000);
+      joinCursorThread.interrupt();
+      sleep(5000);
+    }
+    catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+
+    assertTrue(exceptionHandler.getException() instanceof QueryInterruptedException);
+  }
+
+  public void makeCursorAndAdvance()
+  {
+
+    List<JoinableClause> joinableClauses = ImmutableList.of(
+        factToCountryOnIsoCode(JoinType.LEFT)
+    );
+
+    JoinFilterPreAnalysis joinFilterPreAnalysis = makeDefaultConfigPreAnalysis(
+        null,
+        joinableClauses,
+        VirtualColumns.EMPTY
+    );
+
+    HashJoinSegmentStorageAdapter hashJoinSegmentStorageAdapter = new HashJoinSegmentStorageAdapter(
+        factSegment.asStorageAdapter(),
+        joinableClauses,
+        joinFilterPreAnalysis
+    );
+
+    Cursor cursor = Iterables.getOnlyElement(hashJoinSegmentStorageAdapter.makeCursors(
+        null,
+        Intervals.ETERNITY,
+        VirtualColumns.EMPTY,
+        Granularities.ALL,
+        false,
+        null
+    ).toList());
+
+    ((PostJoinCursor)cursor).setValueMatcher(new ValueMatcher()
+    {
+      @Override
+      public boolean matches(boolean includeUnknown)
+      {
+        return false;
+      }
+
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+
+      }
+    });
+
+    cursor.advance();
+  }
+}
```

Comment:
I think we need to leave a blank line at the end of each file. Not sure though

Reply:
Shall follow. Thanks!